### ISSUE TITLE
HTTP/2 connection coalescing breaks Gateway API

### ISSUE THREAD

**Issue by farcaller** [istio/istio#52853]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

HTTP/2 connection coalescing isn't a new problem. There's an [envoy bug](https://github.com/envoyproxy/envoy/issues/6767) for returning 421, [istio bug](https://github.com/istio/istio/issues/20425), [another istio bug](https://github.com/istio/istio/issues/13589). In fact, it's common enough that it's [documented in the FAQ](https://istio.io/latest/docs/ops/common-problems/network-issues/#404-errors-occur-when-multiple-gateways-configured-with-same-tls-certificate). The workaround mentioned doesn't work for the gateway api, though.

Consider the following scenario. I want to serve `www.farcaller.net` and `farcaller.net` and I want to always redirect the former to the latter. As there are various subdomains, I issue a single cert to handle them all that has `farcaller.net,*.farcaller.net` in its SANs.

Now, with istio gateway API the correct way to set it up is to have a single gateway resource that works for both hostnames, and do the redirection in the virtualservice of this gateway. With the gateway api, though, it's impossible, as the hostname filed only accepts a single hostname. I can have a listener for either farcaller.net or *.farcaller.net.

If I add two listeners, then the envoy config has two filter chains, each with the relevant `filter_chain_match.server_names` set to the hostname. However, this breaks the connection coalescing in case when the client comes to `www.farcaller.net`, gets the cert for both, gets 301 to `farcaller.net` and tries to reuse the connection with `host: farcaller.net` while having SNI `www.farcaller.net`. This matches on the wrong filter chain and the result is 404.

Optionally, one could ignore the hostname field of the gateway spec (because it indeed is optional). The end result is then even worse. Initially it seems that now you have a single filter chain and the actual routing is handled by `route_config`. Now you can reuse the connection with `host: farcaller.net` while having SNI `www.farcaller.net`, and envoy wouldn't care about the latter.

The problem is that it's impossible to serve any other website with a different certificate, because, by removing the hostname, we removed the SNI capability altogether! So even if you add a second `Gateway` for example.com, you will always be served with a certificate for farcaller.net.

TL;DR:

- envoy cannot handle HTTP/2 connection coalescing if there are several filter chains for different SNIs but the same certificate
- gateway API doesn't allow you to declare more than one SNI per filter chain

### Version

```Text
True of istio 1.23.0 envoy 165f7597596f5b4de7025bb635a7b59d1c3fe27e/1.31.1-dev
```

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Comment by keithmattix** [comment]
Not stale

**Comment by farcaller** [comment]
Not stale

**Comment by tolleiv** [comment]
It's 2026 this is still true as of istio 1.29.0

**Comment by keithmattix** [comment]
We should look into this in the next couple of istio releases.

**Comment by keithmattix** [comment]
Did some research; Gateway API's recommendation is to return 421's here (I'm not sure if we implement this in Istio yet). [This doc](https://docs.google.com/document/d/1g_TNN8eOaVDC3xesO9JFdvQbPFdSTHp1vb70TD3-Vrs/edit?tab=t.0) covers the possibility of having multiple hostnames but I don't think that's been implemented in the spec yet
